src/api/source/mvt.js

```diff
@@ -0,0 +1,269 @@
+    }
+
+    _validateInputParams(url, metadata) {
+        validateTemplateURL(url);
+        if (util.isUndefined(metadata)) {
+            throw new CartoValidationError('source', 'metadataRequired');
+        }
+    }
+
+    _buildMetadata(mvtMetadata) {
+        return new Metadata([], mvtMetadata.columns, 0, 0, '');
+    }
+
+    _clone(){
+        return new MVT(this._templateURL, this.metadata);
+    }
+
+    bindLayer(addDataframe, removeDataframe, dataLoadedCallback) {
+        this._addDataframe = addDataframe;
+        this._removeDataframe = removeDataframe;
+        this._dataLoadedCallback = dataLoadedCallback;
+    }
+
+    async requestMetadata(viz) {
+        this._MNS = viz.getMinimumNeededSchema();
+        return this.metadata;
+    }
+
+    requestData(viewport) {
+        const tiles = rsys.rTiles(viewport);
+        this._getTiles(tiles);
+    }
+
+    free() {
+        this.cache.reset();
+        this._oldDataframes = [];
+    }
+
+    _getTiles(tiles) {
+        this._requestGroupID++;
+        var completedTiles = [];
+        var needToComplete = tiles.length;
+        const requestGroupID = this._requestGroupID;
+        tiles.forEach(t => {
+            const { x, y, z } = t;
+            this._getDataframe(x, y, z).then(dataframe => {
+                if (dataframe.empty) {
+                    needToComplete--;
+                } else {
+                    completedTiles.push(dataframe);
+                }
+                if (completedTiles.length == needToComplete && requestGroupID == this._requestGroupID) {
+                    this._oldDataframes.map(d => d.active = false);
+                    completedTiles.map(d => d.active = true);
+                    this._oldDataframes = completedTiles;
+                    this._dataLoadedCallback();
+                }
+            });
+        });
+    }
+
+    _getDataframe(x, y, z) {
+        const id = `${x},${y},${z}`;
+        const c = this.cache.get(id);
+        if (c) {
+            return c;
+        }
+        const promise = this._requestDataframe(x, y, z);
+        this.cache.set(id, promise);
+        return promise;
+    }
+
+    _requestDataframe(x, y, z) {
+        const mvt_extent = 4096;
+
+        return fetch(this._getTileUrl(x, y, z))
+            .then(rawData => rawData.arrayBuffer())
+            .then(response => {
+
+                if (response.byteLength == 0 || response == 'null') {
+                    return { empty: true };
+                }
+                var tile = new VectorTile(new Protobuf(response));
+                const mvtLayer = tile.layers[Object.keys(tile.layers)[0]];
+                var fieldMap = {};
+                const numFields = [];
+                const jsonFields = [];
+                const stringFields = [];
+                this.metadata.columns.map(c => {
+                    const type = c.type;
+                    if (type == 'json') {
+                        jsonFields.push(c.name);
```

Comment:
Probably we should add arrays in the future too

Reply:
Well... This has been discussed a lot, particularly a hook system that allows developers to do pre-processing of the MVTs before passing them to CARTO VL rendering system. But, we decided to don't do it in the near future unless we see that we really need it.